### DISCUSSION TITLE
MvvmDialogs for Mobile

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8887] [Q&A]
[MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/) is working pretty well for Desktop applications. Now that mobile is coming out... what does it mean for the library? I got no experience with mobile development so I'd like your feedback.

This, for example, won't apply for mobile. What can I do instead?

```c#
private static IEnumerable<Window> Windows =>
    (Application.Current?.ApplicationLifetime as IClassicDesktopStyleApplicationLifetime)?.Windows ?? Array.Empty<Window>();

/// <inheritdoc />
public override IView? FindWindowByViewModel(INotifyPropertyChanged viewModel) =>
    Windows.FirstOrDefault(x => ReferenceEquals(viewModel, x.DataContext)).AsWrapper();
```

On mobile, you don't show dialogs, but rather, navigate between pages. Does that page navigation have to be managed by MvvmDialogs or that's something entirely different? There are things like browsing for files and images; not sure how that works on mobile.

Any pointers are welcomed!

**Comment by kekekeks** [comment]
Most likely you will need to implement a navigation stack to be hosted in the main view. We currently don't have a built-in one and have no plans to implement it for 11.0.0 since our schedule is packed as it is.

**Reply by mysteryx93** [reply]
There's no navigation feature built into Avalonia at all?

FluentAvalonia does have a navigation control... Recreating that user control is out of the scope of MvvmDialogs. Is the best bet to rely on FluentAvalonia's navigation control?

*Note that FluentAvalonia isn't yet compatible with v11

The application is basically a View that is constantly visible, with various usercontrols inside a navigation manager.

I'd just need to get that reference to the NavigationManager on the main View, and then navigate to UserControls instead of Windows.

ViewLocator class can bind to either MainView or MainViewMobile depending on whether we run in mobile mode or not... can duplicate the UI, or have MainView contain MainViewMobile usercontrol.

**Comment by mysteryx93** [comment]
How does the mobile "back" button works with Avalonia?

**Reply by maxkatz6** [reply]
It's not handled as for now

**Reply by mysteryx93** [reply]
so for now it's simply hosting a basic single-page window on the iOS and Android device without integration of phone features?

I guess that's the very first step to get stable before doing anything fancier.

In the future; how should I expect this to work, to design accordingly?

**Reply by maxkatz6** [reply]
> so for now it's simply hosting a basic single-page window on the iOS and Android device without integration of phone features?

Yes, but you can access native APIs if you want to implement support for the back button.

> In the future; how should I expect this to work, to design accordingly?

I don't know yet. We don't have plans for the back button support in 11.0, but maybe in 11.x. It could be an event on ApplicationLifetime interface.
